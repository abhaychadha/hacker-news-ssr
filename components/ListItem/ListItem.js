import { useState, useEffect } from 'react';
import { Row, Col } from 'react-bootstrap';
import {StyledRow, StyledCountLabel, StyledArrowUp, StyledInfo, StyledTitle} from './ListItem.styles';
export const ListItem = ({ newsItem, index }) => {
  const [upvoted, setUpvoted] = useState(false);
  const [upvotes, setUpvotes] = useState(newsItem.points);
  useEffect(() => {
    if (upvoted) {
      setUpvotes(upvotes + 1);
    }
  }, [upvoted]);
  return (
    <StyledRow key={newsItem.id} index={index}>
      <Col xl={2} md={2} sm={1} className="text-center"><StyledCountLabel>{newsItem.num_comments}</StyledCountLabel></Col>
      <Col xl={10} md={10} sm={5}><StyledCountLabel upvoted={ upvoted }>{upvotes || 0} </StyledCountLabel>
        <StyledArrowUp onClick={() => setUpvoted(true) } ></StyledArrowUp>
        <StyledTitle>{newsItem.title} </StyledTitle>
        <StyledInfo>{` (${getUrlDomain(newsItem.url)}) by`} </StyledInfo>
        <StyledInfo dark={true}>{newsItem._highlightResult.author.value}</StyledInfo>
        <StyledInfo>{` ${timeSince(new Date(newsItem.created_at))}`}</StyledInfo>
      </Col>
    </StyledRow>);
};

const timeSince = (date) => {
  var seconds = Math.floor((new Date() - date) / 1000);
  var interval = Math.floor(seconds / 31536000);

  if (interval > 1) {
    return interval + ' years';
  }
  interval = Math.floor(seconds / 2592000);
  if (interval > 1) {
    return interval + ' months';
  }
  interval = Math.floor(seconds / 86400);
  if (interval > 1) {
    return interval + ' days';
  }
  interval = Math.floor(seconds / 3600);
  if (interval > 1) {
    return interval + ' hours';
  }
  interval = Math.floor(seconds / 60);
  if (interval > 1) {
    return interval + ' minutes';
  }
  return Math.floor(seconds) + ' seconds';
};

const getUrlDomain = url => url ? url.replace('http://', '').replace('https://', '').split(/[/?#]/)[0] : '';
