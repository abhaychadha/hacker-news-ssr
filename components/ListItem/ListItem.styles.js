import styled from 'styled-components';
import { Row } from 'react-bootstrap';

export const StyledRow = styled(Row)`
  background: ${props => props.index % 2 === 0 ? '#f7f0d6' : '#fdf0be'}
`;
export const StyledCountLabel = styled.span`
  font-size: 14px;
  margin: 4px;
  color: ${props => props.upvoted ? 'red' : '#757575'};
`;

export const StyledTitle = styled.span`
  font-size: 14px;
  color: black;
  margin: 4px;
  font-weight: bold;
`;

export const StyledInfo = styled.span`
  font-size: 12px;
  color: ${props => props.dark ? 'black' : '#757575'};;
`;

export const StyledArrowUp = styled.div`
  width: 0; 
  height: 0; 
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  display: inline-block;
  border-bottom: 5px solid #757575;
  margin: 4px;
  cursor: pointer;
`;
