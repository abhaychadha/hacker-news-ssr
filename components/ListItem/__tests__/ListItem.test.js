import React from 'react';
import { shallow, mount } from 'enzyme';
import { ListItem } from '../ListItem';
import { StyledArrowUp } from '../ListItem.styles';
import initStories from '../../../static/initStories';

describe('ListItem', () => {
  it('List item should render correctly in "debug" mode', () => {
    const component = shallow(<ListItem newsItem = {initStories.hits[0]} index={1} />);
    expect(component).toMatchSnapshot();
  });

  it('upvote arrow is clicked and upvotes and updated', () => {
    // mockTryGetValue.mockReturnValueOnce(true);
    const component = mount(<ListItem newsItem = {initStories.hits[0]} index={1}/>);
    component.find(StyledArrowUp).simulate('click');
    const pointsUpdatedComponent = component.findWhere(n => n.prop('upvoted') === true);
    expect(pointsUpdatedComponent).toHaveLength(1);
  });
});