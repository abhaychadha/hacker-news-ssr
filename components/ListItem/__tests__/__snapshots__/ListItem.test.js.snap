// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem List item should render correctly in "debug" mode 1`] = `
<ListItem__StyledRow
  index={1}
>
  <Col
    className="text-center"
    md={2}
    sm={1}
    xl={2}
  >
    <ListItem__StyledCountLabel>
      813
    </ListItem__StyledCountLabel>
  </Col>
  <Col
    md={10}
    sm={5}
    xl={10}
  >
    <ListItem__StyledCountLabel
      upvoted={false}
    >
      3001
       
    </ListItem__StyledCountLabel>
    <ListItem__StyledArrowUp
      onClick={[Function]}
    />
    <ListItem__StyledTitle>
      macOS High Sierra: Anyone can login as “root” with empty password
       
    </ListItem__StyledTitle>
    <ListItem__StyledInfo>
       (twitter.com) by
       
    </ListItem__StyledInfo>
    <ListItem__StyledInfo
      dark={true}
    >
      vladikoff
    </ListItem__StyledInfo>
    <ListItem__StyledInfo>
       2 years
    </ListItem__StyledInfo>
  </Col>
</ListItem__StyledRow>
`;
